### ISSUE TITLE
Custom types can't be referenced more than once

### ISSUE THREAD

**Issue by Erid** [aws-amplify/amplify-data#495]
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

Not applicable

### Amplify APIs

GraphQL API

### Amplify Version

v6

### Amplify Categories

api

### Backend

Amplify Gen 2

### Environment information

<details>

```
# Put output below this line
  System:
    OS: macOS 15.2
    CPU: (8) arm64 Apple M1 Pro
    Memory: 53.89 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.10.0 - ~/.asdf/installs/nodejs/20.10.0/bin/node
    npm: 10.2.3 - ~/.asdf/plugins/nodejs/shims/npm
    pnpm: 8.15.9 - ~/.asdf/installs/nodejs/20.10.0/bin/pnpm
    bun: 1.0.25 - /opt/homebrew/bin/bun
  Browsers:
    Chrome: 131.0.6778.205
    Edge: 131.0.2903.112
    Safari: 18.2
  npmPackages:
    @aws-amplify/auth: 6.9.0 => 6.9.0 
    @aws-amplify/backend: 1.11.0 => 1.11.0 
    @aws-amplify/backend-cli: 1.4.5 => 1.4.5 
    @aws-sdk/client-cognito-identity-provider: 3.716.0 => 3.716.0 
    @aws-sdk/client-lambda: 3.716.0 => 3.716.0 
    @aws-sdk/client-s3: 3.717.0 => 3.717.0 
    @aws-sdk/s3-request-presigner: 3.717.0 => 3.717.0 
    @empat/core: workspace:* => 1.0.0 
    @empat/eslint-config: workspace:* => 1.0.0 
    @types/aws-lambda: 8.10.146 => 8.10.146 
    aws-amplify: 6.11.0 => 6.11.0 
    aws-cdk: 2.173.2 => 2.173.2 
    aws-cdk-lib: 2.173.2 => 2.173.2 
    constructs: 10.4.2 => 10.4.2 
    dotenv: 16.4.5 => 16.4.5 
    esbuild: 0.24.0 => 0.24.0 
    graphql-request: 7.1.0 => 7.1.0 
    openai: 4.68.4 => 4.68.4 
    tsx: 4.19.1 => 4.19.1 
    typescript: 5.7.2 => 5.7.2 
    vitest: 2.1.3 => 2.1.3 
  npmGlobalPackages:
    @builder.io/micro-agent: 0.1.5
    @githubnext/github-copilot-cli: 0.1.36
    corepack: 0.22.0
    hasura-cli: 2.35.2
    npm: 10.2.3
```

</details>

### Describe the bug

When defining a custom type, if I try to reference it as the return type of a query, it works, however, if I do it on more than one query it throws an error.

### Expected behavior

I would expect the Schema to let me reference the same custom type more than once.

### Reproduction steps

1. Define a custom type, like:
```
CustomType: a.customType({
  id: a.string().required(),
  name: a.string().required(),
}),
```

2. Then reference it to be part of the result of a query, like this:
```
getSomething: a
  .query()
  .arguments({
    arg1: a.string().required(),
  })
  .returns(a.ref("CustomType").required().array().required())
  .handler(a.handler.function(getSomething))
  .authorization((allow) => [allow.group("Admin")]),
```

3. That works just as expected, however, then create another query that returns the same type, and get the following error:
```
api dev: Failed to instantiate data construct
api dev: Caused By: Schema validation failed.\n\nThe directive "@aws_cognito_user_pools" can only be used once at this location.\n\nGraphQL request:259:16\n258 |\n259 | type CustomType @aws_cognito_user_pools @aws_api_key @aws_cognito_user_pools(cognito_groups: ["Admin"])\n    |                ^\n260 | {\n\nGraphQL request:259:53\n258 |\n259 | type CustomType @aws_cognito_user_pools @aws_api_key @aws_cognito_user_pools(cognito_groups: ["Admin"])\n    |                                                     ^\n260 | {
api dev: Resolution: See the underlying error message for more details.
api dev: [Sandbox] Watching for file changes...
```

**Comment by chrisbonifacio** [comment]
Hi @Erid, just wanted to follow up and see if you are still blocked on this issue. If you need more time to provide a reproducible app, totally understand! I did try downgrading to `1.11.0` in my local project as well but did not experience the error on deployment.

We also thought maybe the conflict in auth directives might've been caused by having the same auth rule on the schema level so I tried that but alas, the schema still compiled and deployed ☹️ 

```ts
const schema = a
  .schema({
    CustomType: a.customType({
      id: a.string().required(),
      name: a.string().required(),
    }),
    getSomething: a
      .query()
      .arguments({
        arg1: a.string().required(),
      })
      .returns(a.ref("CustomType").required().array().required())
      .handler(a.handler.function(getSomething))
      .authorization((allow) => [allow.group("Admin")]),
    getSomething2: a
      .query()
      .arguments({
        arg1: a.string().required(),
      })
      .returns(a.ref("CustomType").required().array().required())
      .handler(a.handler.function(getSomething))
      .authorization((allow) => [allow.group("Admin")]),
  })
  .authorization((allow) => [allow.group("Admin")]);
```

**Comment by Erid** [comment]
Thanks for the follow up @chrisbonifacio! I need some more time to repro, it's still happening to me, I "worked around it" by duplicating the type for now; but this week I didn't have the time to create the app to repro and I'm away from my computer for a few days now.

I'll have it for you next week! I need to find the root cause 😅

**Comment by chrisbonifacio** [comment]
Hi 👋 Closing this as we have not heard back from you. If you are still experiencing this issue and in need of assistance, please feel free to comment and provide us with any information previously requested by our team members so we can re-open this issue and be better able to assist you.

Thank you!

**Comment by frederikbendel** [comment]
i am experiencing this as well.

for reproducing you may try the following:
first deploy with only one group access
then deploy with group access for an additional other group. 

e.g.
1. [allow.groups(["Admin")]),

2. [allow.groups(["Admin","User")]),

**Comment by Erid** [comment]
Thanks @frederikbendel! Just yesterday I had the issue again and realized that that's why I was getting the problem, and it makes me more confident with someone else reproducing it, too. So to reproduce, you would need to do the following:

```typescript
  CustomType: a.customType({
    id: a.string().required(),
    name: a.string().required(),
  }),
  getSomething: a
    .query()
    .arguments({
      arg1: a.string().required(),
    })
    .returns(a.ref("CustomType").required().array().required())
    .handler(a.handler.function(getSomething))
    .authorization((allow) => [allow.group("Admin")]),
  getSomething2: a
    .query()
    .arguments({
      arg1: a.string().required(),
    })
    .returns(a.ref("CustomType").required().array().required())
    .handler(a.handler.function(getSomething))
    .authorization((allow) => [allow.groups(["Admin", "User"])]),
```

I'm sorry @chrisbonifacio, I couldn't go back to you earlier... But I hope now you can properly identify this. I'd really hope you can re-open it, since I've come to the point where I need to create multiple custom types that are identical, and it becomes hard to follow if those types rely on another custom type.

**Comment by chrisbonifacio** [comment]
Thank you @Erid & @frederikbendel for clarifying the reproduction step!

Repro App: https://github.com/chrisbonifacio/issue-495

I was able to reproduce the issue consistently in two ways:

Auth Rule
```ts
    .authorization((allow) => [allow.groups(["Admin", "User"])]),
```

Error

![Image](https://github.com/user-attachments/assets/f7ef8135-be37-47a0-94b9-e14e1b333354)


Auth Rule
```ts
    .authorization((allow) => [allow.group("Admin"), allow.group("User")]),
```

Error

![Image](https://github.com/user-attachments/assets/b17261de-4bb4-4d38-8a25-af484c812b9f)

**Comment by Erid** [comment]
Hi @stocaaro!! Thanks for the fix, but I have bad news... I tried and it still happens, but you need to have more than 2 references, like the following:
```typescript
  CustomType: a.customType({
    id: a.string().required(),
    name: a.string().required(),
  }),
  getSomething: a
    .query()
    .arguments({
      arg1: a.string().required(),
    })
    .returns(a.ref("CustomType").required().array().required())
    .handler(a.handler.function(getSomething))
    .authorization((allow) => [allow.group("Admin")]),
  getSomething2: a
    .query()
    .arguments({
      arg1: a.string().required(),
    })
    .returns(a.ref("CustomType").required().array().required())
    .handler(a.handler.function(getSomething))
    .authorization((allow) => [allow.groups(["Admin", "User"])]),
  getSomething3: a
    .query()
    .arguments({
      arg1: a.string().required(),
    })
    .returns(a.ref("CustomType").required().array().required())
    .handler(a.handler.function(getSomething))
    .authorization((allow) => [allow.groups(["Admin", "User"])]),
```

I was excited to test it out, and in my case I had the situation where `getSomething2` and `getSomething3` where sharing a new custom type (identical to the first), however, when I changed them I got the same error, and then I made it so only `getSomething2` used the same custom type and it worked, but `getSomething3` still needed to use a different custom type :/

cc @chrisbonifacio

**Comment by stocaaro** [comment]
Hi @Erid ,

Appreciate your patience with this. I thought I had it fixed 😅 .  Is the error message that your getting the same as before?


Thanks,
Aaron

**Comment by Erid** [comment]
Thanks @stocaaro, yes, it's the same error:
```
The directive "@aws_cognito_user_pools" can only be used once at this location.
```

Another thing worth noting is that I the custom type I use references another custom type. I did remove the reference to the second custom type and I still got the error, just in less places.

Something like:
```typescript
  Author: a.customType({
    id: a.string().required(),
    name: a.string().required(),
  }),
  CustomType2: a.customType({
    id: a.string().required(),
    name: a.string().required(),
    author: a.ref("Author"),
  }),
```

I made sure I had the `@aws-amplify/data-schema` at `1.20.2`. Also, my schema has plenty of other custom types, models, queries and mutations, just in case that makes a difference.

**Comment by stocaaro** [comment]
I think I found the issue with my original solution and I have a fix out for review. I'll post another update when I get it released.

Thanks,
Aaron

**Comment by stocaaro** [comment]
I put together a [further fix](https://github.com/aws-amplify/amplify-data/pull/574/) for this last week. My prior testing has focused on getting the sandbox build to work and have the expected definition, however in further testing we have uncovered runtime issues where the library code that handles the runtime behavior has issues.

On one hand my fix solves the immediate issue without introducing any new issues directly. Indirectly, my fix reveals a class of runtime issues that we don't have answers for right now.

This work is in a holding pattern while we evaluate how to solve the runtime issues with these recursive times.

**Comment by stocaaro** [comment]
Hello @jsanders5, This fix has been released. Can you update to the latest version of `@aws-amplify/data-schema` and confirm that the fix works for your application?

Thanks,
Aaron
